// Executor Service - Order execution with Alpaca paper trading integration
// Part of BLOASIS Phase 1 infrastructure for trade execution
//
// This service handles order execution after Risk Committee approval:
// - Submits orders to Alpaca paper trading
// - Supports market, limit, and bracket orders
// - Tracks order status and fills
// - Publishes execution events to Redpanda
//
// Dependencies:
// - Risk Committee Service: Approval verification
// - Redis: Order tracking and risk approval cache
//
// Order Status Flow:
// - pending: Order created, not yet submitted
// - submitted: Order submitted to Alpaca
// - filled: Order fully executed
// - partially_filled: Order partially executed
// - cancelled: Order cancelled
// - rejected: Order rejected by broker
// - expired: Order expired
//
// Note: This service is for paper trading only in Phase 1.

syntax = "proto3";

package bloasis.executor;

import "google/api/annotations.proto";

// ExecutorService handles order execution through Alpaca paper trading
service ExecutorService {
  // ExecuteOrder submits an order for execution
  // Requires prior Risk Committee approval
  rpc ExecuteOrder(ExecuteOrderRequest) returns (ExecuteOrderResponse) {
    option (google.api.http) = {
      post: "/v1/executor/order"
      body: "*"
    };
  }

  // GetOrderStatus retrieves current status of an order
  rpc GetOrderStatus(GetOrderStatusRequest) returns (GetOrderStatusResponse) {
    option (google.api.http) = {
      get: "/v1/executor/order/{order_id}"
    };
  }

  // CancelOrder cancels an open order
  rpc CancelOrder(CancelOrderRequest) returns (CancelOrderResponse) {
    option (google.api.http) = {
      delete: "/v1/executor/order/{order_id}"
    };
  }

  // GetAccount retrieves account information
  rpc GetAccount(GetAccountRequest) returns (GetAccountResponse) {
    option (google.api.http) = {
      get: "/v1/executor/account"
    };
  }
}

// ExecuteOrderRequest - Request to execute an order
message ExecuteOrderRequest {
  // User identifier
  string user_id = 1;
  // Stock ticker symbol (e.g., "AAPL", "GOOGL")
  string symbol = 2;
  // Order side: "buy" or "sell"
  string side = 3;
  // Quantity to trade
  double qty = 4;
  // Order type: "market", "limit", "bracket"
  string order_type = 5;
  // Limit price (required for limit orders)
  double limit_price = 6;
  // Stop loss price (for bracket orders)
  double stop_loss = 7;
  // Take profit price (for bracket orders)
  double take_profit = 8;
  // Risk approval ID from Risk Committee (required)
  string risk_approval_id = 9;
}

// ExecuteOrderResponse - Order execution result
message ExecuteOrderResponse {
  // Whether the order was submitted successfully
  bool success = 1;
  // Alpaca order ID
  string order_id = 2;
  // Client-generated order ID (format: bloasis-xxxxxxxx)
  string client_order_id = 3;
  // Order status (pending, submitted, filled, rejected, etc.)
  string status = 4;
  // Error message if failed
  string error_message = 5;
}

// GetOrderStatusRequest - Request for order status
message GetOrderStatusRequest {
  // Alpaca order ID
  string order_id = 1;
}

// GetOrderStatusResponse - Order status details
message GetOrderStatusResponse {
  // Alpaca order ID
  string order_id = 1;
  // Current status
  string status = 2;
  // Quantity filled
  double filled_qty = 3;
  // Average fill price
  double filled_avg_price = 4;
  // Submission timestamp (ISO 8601)
  string submitted_at = 5;
  // Fill timestamp (ISO 8601, if filled)
  string filled_at = 6;
}

// CancelOrderRequest - Request to cancel an order
message CancelOrderRequest {
  // User identifier
  string user_id = 1;
  // Alpaca order ID to cancel
  string order_id = 2;
}

// CancelOrderResponse - Cancellation result
message CancelOrderResponse {
  // Whether cancellation succeeded
  bool success = 1;
  // Status message
  string message = 2;
}

// GetAccountRequest - Request for account info
message GetAccountRequest {
  // User identifier
  string user_id = 1;
}

// GetAccountResponse - Account information from Alpaca
message GetAccountResponse {
  // Available cash
  double cash = 1;
  // Buying power
  double buying_power = 2;
  // Total portfolio value
  double portfolio_value = 3;
  // Account equity
  double equity = 4;
}
